pkg/controllers/interruption/controller.go

```diff
@@ -0,0 +1,275 @@
+}
+
+func NewController(kubeClient client.Client, clk clock.Clock, recorder events.Recorder, cluster *state.Cluster,
+	sqsProvider *providers.SQS, unavailableOfferingsCache *awscache.UnavailableOfferings) *Controller {
+
+	return &Controller{
+		kubeClient:                kubeClient,
+		clk:                       clk,
+		cluster:                   cluster,
+		recorder:                  recorder,
+		provider:                  sqsProvider,
+		unavailableOfferingsCache: unavailableOfferingsCache,
+		parser:                    NewEventParser(DefaultParsers...),
+	}
+}
+
+func (c *Controller) Reconcile(ctx context.Context, _ reconcile.Request) (reconcile.Result, error) {
+	defer metrics.Measure(reconcileDuration)()
+	list := &v1alpha1.AWSNodeTemplateList{}
+	if err := c.kubeClient.List(ctx, list); err != nil {
+		return reconcile.Result{}, fmt.Errorf("listing node templates, %w", err)
+	}
+
+	if awssettings.FromContext(ctx).EnableInterruptionHandling && len(list.Items) > 0 {
+		active.Set(1)
+		sqsMessages, err := c.provider.GetSQSMessages(ctx)
+		if err != nil {
+			return reconcile.Result{}, fmt.Errorf("getting messages from queue, %w", err)
+		}
+		if len(sqsMessages) == 0 {
+			return reconcile.Result{}, nil
+		}
+		instanceIDMap := c.makeInstanceIDMap()
+		errs := make([]error, len(sqsMessages))
+		workqueue.ParallelizeUntil(ctx, 10, len(sqsMessages), func(i int) {
+			errs[i] = c.handleMessage(ctx, instanceIDMap, sqsMessages[i])
+		})
+		return reconcile.Result{}, multierr.Combine(errs...)
+	}
+	active.Set(0)
+	return reconcile.Result{RequeueAfter: time.Second * 10}, nil
+}
+
+func (c *Controller) Builder(_ context.Context, m manager.Manager) operatorcontroller.Builder {
+	return operatorcontroller.NewSingletonManagedBy(m).
+		Named("interruption")
+}
+
+func (c *Controller) LivenessProbe(_ *http.Request) error {
+	return nil
+}
+
+// handleMessage gets the node names of the instances involved in the queue message and takes the
+// assigned action on the instances based on the message event
+func (c *Controller) handleMessage(ctx context.Context, instanceIDMap map[string]*v1.Node, raw *sqsapi.Message) error {
+	// No message to parse in this case
+	if raw == nil || raw.Body == nil {
+		return nil
+	}
+	msg, err := c.parser.Parse(*raw.Body)
+	if err != nil {
+		// In the scenario where we can't parse the message, we log that we have an error and then are
+		// forced to just delete the message from the queue
+		logging.FromContext(ctx).Errorf("parsing sqs message, %v", err)
+		err = c.provider.DeleteSQSMessage(ctx, raw)
+		if err != nil {
+			return fmt.Errorf("failed to delete message from queue, %w", err)
+		}
+		deletedMessages.Inc()
+		return nil
+	}
+	ctx = logging.WithLogger(ctx, logging.FromContext(ctx).With("event", msg.Kind()))
+
+	nodes := getInvolvedNodes(msg.EC2InstanceIDs(), instanceIDMap)
+	// There's no action to take here since the event doesn't pertain to any of our instances
+	if len(nodes) == 0 {
```

Comment:
optional: I think a lot of code would collapse by using implicit zero looping. i.e, loop over a list of things, which implicitly doesn't take an action if the list is emtpy
```
for _, instanceId := msg.EC2InstanceID {
   node, ok := instanceIDMap[instanceIdMap]
   if !ok {
       continue
   }
}
// Always do this if handled.
c.provider.DeleteSQSMessage
```